lib/factree/fact_source.rb

```diff
@@ -0,0 +1,103 @@
+module Factree
+  # Mixin to help define fact source classes.
+  module FactSource
+    UndefinedFactError = Class.new(StandardError)
+    UnknownFactError = Class.new(StandardError)
+
+    def self.included(base)
+      base.extend(ClassMethods)
+    end
+
+    module ClassMethods
+      # Defines a named fact along with a method-like block used to compute the fact's value.
+      #
+      # @param [Symbol] fact_name The new fact's name
+      def def_fact(fact_name, &block)
+        all_fact_procs = fact_procs.merge(fact_name => block).freeze
+        class_variable_set(:@@_fact_procs, all_fact_procs)
+        fact_name
+      end
+
+      # @return [Array<Symbol>] Names for all of the defined facts, whether their values are known or not.
+      def fact_names
+        fact_procs.keys.freeze
+      end
+
+      # @api private
+      def fact_procs
+        class_variable_defined?(:@@_fact_procs) ?
+          class_variable_get(:@@_fact_procs) :
+          {}.freeze
+      end
+
+      # @return [Boolean] True if the fact has been defined (using {FactSource.def_fact})
+      def defined?(fact_name)
+        fact_procs.include? fact_name
```

Comment:
`include?` is the most common way of checking inclusion in collections, and it's the term used by [most](http://test-unit.github.io/test-unit/en/Test/Unit/Assertions.html#assert_include-instance_method) [testing](http://ruby-doc.org/stdlib-2.0.0/libdoc/minitest/rdoc/MiniTest/Expectations.html#method-i-must_include) [frameworks](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/include-matcher).